nexus/db-queries/src/db/datastore/sled.rs

```diff
@@ -291,30 +347,197 @@ impl DataStore {
+                    // policy = 'fail'.
+                    // - A,P=Allow: Same as above, but with policy = 'allow'.
+                    //
+                    // ## Affinity: Apply Rules
+                    //
+                    // - Targets := All viable sleds for instance placement
+                    // - Banned := AA,P=Fail
+                    // - Required := A,P=Fail
+                    // - if Required.len() > 1: Fail (too many constraints).
+                    // - if Required.len() == 1...
+                    //   - ... if the entry exists in the "Banned" set: Fail
+                    //     (contradicting constraints 'Banned' + 'Required')
+                    //   - ... if the entry does not exist in "Targets": Fail
+                    //     ('Required' constraint not satisfiable)
+                    //   - ... if the entry does not exist in "Banned": Use it.
+                    //
+                    // If we have not yet picked a target, we can filter the
+                    // set of targets to ignore "banned" sleds, and then apply
+                    // preferences.
+                    //
+                    // - Targets := Targets \ Banned
+                    //
+                    // ## Affinity: Apply Preferences
+                    //
+                    // - Preferred := Targets ∩ A,P=Allow
+                    // - Unpreferred := Targets ∩ AA,P=Allow
+                    // - Neither := Preferred ∩ Unpreferred
+                    // - Preferred := Preferred \ Neither
+                    // - Unpreferred := Unpreferred \ Neither
+                    // - If Preferred isn't empty, pick a target from it.
+                    // - Targets := Targets \ Unpreferred
+                    // - If Targets isn't empty, pick a target from it.
+                    // - If Unpreferred isn't empty, pick a target from it.
+                    // - Fail, no targets are available.
+
+                    let mut targets: HashSet<_> = sled_targets.into_iter().collect();
+
+                    let banned = anti_affinity_sleds.iter().filter_map(|(policy, id)| {
+                        if *policy == AffinityPolicy::Fail {
+                            Some(*id)
+                        } else {
+                            None
+                        }
+                    }).collect::<HashSet<_>>();
+                    let required = affinity_sleds.iter().filter_map(|(policy, id)| {
+                        if *policy == AffinityPolicy::Fail {
+                            Some(*id)
+                        } else {
+                            None
+                        }
+                    }).collect::<HashSet<_>>();
+
+                    if !banned.is_empty() {
+                        info!(
+                            opctx.log,
+                            "affinity policy prohibits placement on {} sleds", banned.len();
```

Comment:
nitpicky (sorry): i might have this say
```suggestion
                            "anti-affinity policy prohibits placement on {} sleds", banned.len();
```
because i imagine someone grepping the logs for the string "anti-affinity" to find stuff about anti-affinity policies being applied?

Reply:
Sounds good, updated